### ISSUE TITLE
Add ability to specify buildId value in publications

### ISSUE THREAD

**Issue by liutikas** [gradle/gradle#31218]
### Expected Behavior

Add a method `publication.withBuildIdentifier("MYBUILDID")` for creating Gradle module metadata that contains the specified Build ID.

```
  "createdBy": {
    "gradle": {
      "buildId": "MYBUILDID"
    }
  },
```

### Current Behavior (optional)

`publication.withBuildIdentifier()` enables creating of 
```
  "createdBy": {
    "gradle": {
      "buildId": "7naq2pyxu5d3jixk7cvxso7tpu"
    }
  },
```

but it does not allow you to specify the value of it, you just get a random hash.

So today, we do [a really ugly workaround that rewrites this id in a hacky way](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:buildSrc/private/src/main/kotlin/androidx/build/MavenUploadHelper.kt;l=274)

### Context

Our build system has a concept of a build id and we want to use it to tag our artifacts based on that.

**Comment by ov7a** [comment]
This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by big-guy** [comment]
I don't think we want to make this configurable.  If anything, we may just get rid of buildId from GMM entirely.

**Comment by liutikas** [comment]
Do you have a recommendation on where to put the metadata such a build into
our artifacts?

**Comment by big-guy** [comment]
Tradeoffs abound.  It depends on how important this information is.

1. If you want to be able to identify the build based purely on the artifact, you can embed this in the artifact. This may impact caching, but there are ways around that: https://docs.gradle.org/current/userguide/incremental_build.html#sec:configure_input_normalization
2. If you only care about querying this later, some Artifact repositories (like Artifactory) have APIs to capture this information when publishing.
3. You could capture build information as a separate artifact to publish.

The original idea for capturing the "buildId" was to provide metadata to other systems that would identify the Gradle build invocation. If this was configurable, we wouldn't be able to rely on it representing a Gradle build invocation (since it could be something like the Jenkins build number). 

We're not going to be able to consider making changes to this before 9.0.
